### PR TITLE
481 feedback

### LINKED ISSUE #492: Shallow copy of OmegaConf containers does not work correctly
OmegaConf containers has a reference to the parent node.
This is to provide features like interpolations and the ability to provide the full key of a node (including the parent keys) in errors.

This means that there is a bi-birectional links between a container (DictConfig, ListConfig) and a node (DictConfig, ListConfig, AnyNode, IntNode etc).

Consider:
```
a:
  b:
    c: 10
```
The links are as follows:
```
a -> b, b -> a
b -> c, c -> b
```
If we copy a with a shallow copy, we end up with:

```
a -> b
b -> c, c -> b
# new links
a1 -> b,  b -> a1
```
Since b is the same object (shallow copy!), it can only have one parent. When we set the parent to a1, the parent is no longer a.
this is breaking the data model. if b (or one of it's children) is an interpolation, it's value might be changed now by changes to the a1 config tree.

One may think that we can solve it by copying b as well:
```
a -> b, b -> a
b -> c

a1 -> b1, b1 -> a1
b1 -> c, c -> b1
```
But now we have the same problem with c. 

The only real solution is to perform a deepcopy instead of a shallow copy.

### PR DESCRIPTION
**Pull request by omry**
Updates per the feedback on #481 

Closes #492

### CODE REVIEW

File: tests/test_base_config.py [hunk 1]
```diff
@@ -618,14 +637,34 @@ def test_copy_with_interpolation(
         cp[interpolated_key] = "XXX"
         assert cp[interpolated_key] == cp[interpolating_key]
 
-    def test_list_copy_is_shallow(self, copy_method: Any) -> None:
+    def test_list_shallow_copy_is_deepcopy(self, copy_method: Any) -> None:
```
Comment (odelalleau) [inline tests/test_base_config.py:640]: Should there be a similar test for dicts?

File: omegaconf/dictconfig.py [hunk 2]
```diff
@@ -130,30 +131,13 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "DictConfig":
             content_copy = src_content
 
         res.__dict__["_content"] = content_copy
-
+        # parent is retained, but not copied
+        res.__dict__["_parent"] = self.__dict__["_parent"]
         return res
 
     def __copy__(self) -> "DictConfig":
```
Comment (odelalleau) [inline omegaconf/dictconfig.py:138]: Move `__copy__` to the `BaseContainer` class? Or do we need it here for typing purpose?

**Review by odelalleau — CHANGES_REQUESTED** [review]
Looks good, just some minor suggestions

File: tests/test_base_config.py [hunk 3]
```diff
@@ -324,6 +324,25 @@ def test_set_flags() -> None:
...
+    assert nc._flags_cache is None
+    assert nc._get_flag("foo")
+    assert nc._flags_cache == {"foo": True}
+
+    cfg2 = OmegaConf.create(flags={"no_deepcopy_set_nodes": no_deepcopy_set_nodes})
+    cfg2._set_flag("foo", False)
+    cfg2.c = cfg._get_node("c")
+    nc = cfg2._get_node("c")
+    assert nc is not None
+    assert not nc._get_flag("foo")
```
[the remark this answers is not in the document]
Comment (omry) [reply tests/test_base_config.py:343]: flags are Optional[bool], so I don't see a point in testing with anything else.

File: omegaconf/nodes.py [hunk 4]
```diff
@@ -72,12 +72,15 @@ def __ne__(self, other: Any) -> bool:
     def __hash__(self) -> int:
         return hash(self._val)
 
-    def _deepcopy_impl(self, res: Any, memo: Optional[Dict[int, Any]] = {}) -> None:
+    def _deepcopy_impl(self, res: Any, memo: Optional[Dict[int, Any]]) -> None:
         res.__dict__["_metadata"] = copy.deepcopy(self._metadata, memo=memo)
-        res.__dict__["_parent"] = copy.deepcopy(self._parent, memo=memo)
+        # TODO retain parent without copying
```
[the remark this answers is not in the document]
Comment (omry) [reply omegaconf/nodes.py:77]: oops.

File: tests/test_base_config.py (hunk 1 above)
Comment (omry) [reply tests/test_base_config.py:640]: Yeah, was a bit lazy.
I think we should probably reorganize all the copy tests and put them in a single test file. it's quiet a mess now.

**Comment by omry** [comment]
Updated per feedback, thanks.

File: tests/test_base_config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (odelalleau) [reply tests/test_base_config.py]: Oops my bad, the `is` version that you implemented is indeed the correct thing (it only really mattered for the `False` value) => resolving

**Review by odelalleau — APPROVED** [review]
All good, thanks!

**Pull request merged**